blaze-core/src/main/scala/org/http4s/blazecore/util/EntityBodyWriter.scala

```diff
@@ -51,18 +51,25 @@ private[http4s] trait EntityBodyWriter[F[_]] {
   /** Called in the event of an Await failure to alert the pipeline to cleanup */
   protected def exceptionFlush(): Future[Unit] = FutureUnit
 
-  /** Creates an effect that writes the contents of the EntityBody to the output.
-    * The writeBodyEnd triggers if there are no exceptions, and the result will
-    * be the result of the writeEnd call.
+  /** Creates an effect that writes the contents of the [[Entity]] to the output.
     *
-    * @param p EntityBody to write out
+    * @param entity an [[Entity]] which body to write out
     * @return the Task which when run will unwind the Process
     */
-  def writeEntityBody(p: EntityBody[F]): F[Boolean] = {
-    val writeBody: F[Unit] = writePipe(p).compile.drain
-    val writeBodyEnd: F[Boolean] = fromFutureNoShift(F.delay(writeEnd(Chunk.empty)))
-    writeBody *> writeBodyEnd
-  }
+  def writeEntityBody(entity: Entity[F]): F[Boolean] =
+    entity match {
+      case Entity.Default(body, _) =>
+        val writeBody: F[Unit] = writePipe(body).compile.drain
+        val writeBodyEnd: F[Boolean] = fromFutureNoShift(F.delay(writeEnd(Chunk.empty)))
+        writeBody *> writeBodyEnd
+
+      case Entity.Strict(chunk) =>
+        fromFutureNoShift(F.delay(writeBodyChunk(chunk, flush = false))) *>
+          fromFutureNoShift(F.delay(writeEnd(Chunk.empty)))
```

Comment:
Can we just `writeEnd(chunk)` instead of doing two writes?

Reply:
I feel that would mean behavior changes then. We can safely do so for the `BodylessWriter`. But I'm not sure about the cases of other writers, say `CachingChunkWriter`. Nevertheless, I could misapprehend things and would be grateful for clarification.

Reply:
I looked, and it's not going to be a big difference on `CachingChunkWriter`.  They're both ultimately going to go through `writeBodyChunk`, and then a flatMap, and then write the trailer.  Similarly with `FushingChunkWriter`.

I would say it's better to call `writeEnd` when you know it's the last chunk.  That's why it takes a chunk.  And a writer implementation may or may not be optimized for that knowledge.  We just haven't ever known it was the last chunk since the days of scalaz-stream.

Reply:
I've tried to change `writeEntityBody` on case `Entity.Strict` and got a bunch of failed tests within `blaze-server`:
```
org.http4s.blaze.server.BlazeServerSuite
org.http4s.blaze.server.BlazeServerMtlsSpec
org.http4s.blaze.server.Http1ServerStageSpec
```
Roughly 1.5 dozen of tests. 
```scala
def writeEntityBody(entity: Entity[F]): F[Boolean] =
  entity match {
    case Entity.Default(body, _) =>
      val writeBody: F[Unit] = writePipe(body).compile.drain
      val writeBodyEnd: F[Boolean] = fromFutureNoShift(F.delay(writeEnd(Chunk.empty)))
      writeBody *> writeBodyEnd

    case Entity.Strict(_) | Entity.Empty =>
      fromFutureNoShift(F.delay(writeEnd(Chunk.empty)))
  }
```

Reply:
I think you would need:

```scala
case Entity.Strict(chunk) =>
  fromFutureNoShift(F.delay(writeEnd(chunk)))
```

It can't share with `Entity.Empty`, because that last chunk needs to be passed to `writeEnd`.

Reply:
Darn. Now I see what did you mean. Locally works fine, all tests passed. Thanks!